### PR TITLE
Implement a subset of IAM resources

### PR DESCRIPTION
**Pull request by bitglue**
Implements support for IAM users, user policies, and access keys. This
is only a subset of what IAM can do (notably missing: roles and instance
profiles and associated policies), but it's a start.

Makes a dent in #28.

The user policies and access keys are implemented as their own resources, rather than being sub-trees in the `aws_iam_user`, to support use cases where one module defines a user, and other modules add policies or keys.

### CODE REVIEW

**Comment by mitchellh** [comment]
This actually all looks really good. Can you add acceptance tests? Once that is in, I'm happy to merge. Thanks for doing docs, as well.

**Comment by bitglue** [comment]
I'll give it a shot. Might be some days or weeks.

**Comment by radeksimko** [comment]
:+1: 
Except acceptance tests, it may also need links in the website sidebar:

```diff
diff --git a/website/source/layouts/aws.erb b/website/source/layouts/aws.erb
index 030192d..e14ba79 100644
--- a/website/source/layouts/aws.erb
+++ b/website/source/layouts/aws.erb
@@ -41,6 +41,18 @@
                                        <a href="/docs/providers/aws/r/elb.html">aws_elb</a>
                     </li>
 
+                    <li<%= sidebar_current("docs-aws-resource-iam=access-key") %>>
+                    <a href="/docs/providers/aws/r/iam_access_key.html">aws_iam_access_key</a>
+                    </li>
+
+                    <li<%= sidebar_current("docs-aws-resource-iam-user") %>>
+                    <a href="/docs/providers/aws/r/iam_user.html">aws_iam_user</a>
+                    </li>
+
+                    <li<%= sidebar_current("docs-aws-resource-iam-user-policy") %>>
+                    <a href="/docs/providers/aws/r/iam_user_policy.html">aws_iam_user_policy</a>
+                    </li>
+
                     <li<%= sidebar_current("docs-aws-resource-instance") %>>
                                        <a href="/docs/providers/aws/r/instance.html">aws_instance</a>
                                        </li>
```

and it would be nice to have `aws_iam_role` resource as well.
http://docs.aws.amazon.com/IAM/latest/APIReference/API_CreateRole.html

Most importantly some more significant changes will need to be done to make it work with the new `aws-sdk-go` library which is now being used for almost all AWS resources (no more `goamz`).

**Comment by bitglue** [comment]
Agreed, @radeksimko. Thanks for the feedback. I will tackle those issues eventually. Currently I'm focused on delivering an internal demo and proof of concept, but when that's done (and successful, I hope), I'll be making a finishing pass over this and many other things. Still some ways out on my priority list, however.

**Comment by bitglue** [comment]
I've updated the website as @radeksimko suggested, and updated to use  aws-sdk-go. Still needs some work on the rough edges, and acceptance tests.

**Comment by bitglue** [comment]
@radeksimko roles and instance profiles are implemented now.

I did some horrible copy&paste to implement groups/users/roles, which really are almost exactly the same file. I'll have to do some thinking on how to clean that up.

Also, I didn't implement IAM managed policies because the vendored AWS SDK didn't support them. But I see now that the current version from aws-labs which does support policies is in use in places -- is migrating to that the goal?

**Comment by radeksimko** [comment]
> But I see now that the current version from aws-labs which does support policies is in use in places -- is migrating to that the goal?

I believe it is, @catsby started migrating all existing AWS resources over back to the vendored AWS SDK. I would probably hold on until all the resources are migrated and "workarounds" (like import aliases) removed, then I'd update the code so it uses the vendored version.

**Comment by catsby** [comment]
I'm in the process of migrating to the upstream [awslabs](https://github.com/awslabs) version of `aws-sdk-go`, and retiring our fork. It is a bit of an _as-I-can_ process, while I try to stay on top of bug fixes in the AWS provider as well. 

@bitglue to my knowledge, RDS is the only resource currently using the `IAM` connection I added in #1292. If it's helpful I can convert RDS next, so you can convert/update this PR to use the upstream library and carry on with IAM managed policies.

**Comment by fromonesrc** [comment]
@catsby is there anything you could use some help with in the awslabs migration?

**Comment by catsby** [comment]
@fromonesrc I created https://github.com/hashicorp/terraform/issues/1488 to track the migrations

**Comment by bitglue** [comment]
On Apr 10, 2015 9:57 AM, "Clint" <notifications@github.com> wrote:
> @bitglue to my knowledge, RDS is the only resource currently using the
IAM connection I added in #1292. If it's helpful I can convert RDS next, so
you can convert/update this PR to use the upstream library and carry on
with IAM managed policies.

That would be super great.

**Comment by catsby** [comment]
@bitglue I opened #1510 for converting RDS, and upgrading the IAM connection. I suppose the latter wasn't strictly needed, but I wanted to rid the resource of `hashicorp/aws-sdk-go`. Only RDS instance is converted there; parameter group and subnet to follow.

**Comment by bitglue** [comment]
@catsby Thanks for doing that. I've rebased to the current master so this should now work with the awslabs SDK.

**Comment by johnrengelman** [comment]
This is the biggest missing piece for me right now with managing our environment. +1 for the feature.

**Comment by johnrengelman** [comment]
I've added a number of basic acceptance tests on top of these commits here: https://github.com/hashicorp/terraform/pull/1591. Hopefully this gets us to a point where we can merge this in.

**Comment by bitglue** [comment]
If we can wait a day or two, I'm adding a resource for managed policies. And I do have a clean account, so I'll give @johnrengelman's tests a run.

**Comment by bitglue** [comment]
@mitchellh Acceptance tests added by @johnrengelman just needed a few tweaks to work. They all pass for me. Requesting a review and merge.

**Comment by johnrengelman** [comment]
Looks like we are still missing documentation for many of the resources.

**Comment by bitglue** [comment]
:sob: too exhausted to work on that now. So much copy&paste...

**Comment by johnrengelman** [comment]
I hear ya :) I can't do it right now either. Perhaps they'll merge it and we can add the docs later :)

**Comment by bitglue** [comment]
Or someone will write a documentation generator. There is a description field in the schema which I think currently does exactly nothing.

**Comment by ctiwald** [comment]
I really really really want this feature, so I've started writing docs in a commit based on this pull request located [here.](https://github.com/ctiwald/terraform/commit/9be4e3f895e14bd9b841f657fce8b5fe66117bc9)

I'm 100% fine with you all just copy/pasting directly out of this commit if you guys think the docs are valid. You should probably read them over as I add them because I'm just following the existing docs and modifying according to the structs I see in these commits. To put it another way: It might not be pretty, but it'll probably get you there.

**Comment by bitglue** [comment]
@ctiwald thanks. I've pulled your commit into my branch so it's included in the PR. Just ping me if you want me to update, or you can open a PR against [my branch](https://github.com/bitglue/terraform/tree/iam).

**Comment by ctiwald** [comment]
Will do. In particular I'm not really sure what is and is not exported so I'd definitely double check that.

File: builtin/providers/aws/provider.go [hunk 1]
```diff
@@ -52,6 +52,15 @@ func Provider() terraform.ResourceProvider {
...
 			"aws_elb":                          resourceAwsElb(),
+			"aws_iam_access_key":               resourceAwsIamAccessKey(),
+			"aws_iam_group_policy":             resourceAwsIamGroupPolicy(),
+			"aws_iam_group":                    resourceAwsIamGroup(),
+			"aws_iam_instance_profile":         resourceAwsIamInstanceProfile(),
+			"aws_iam_policy":                   resourceAwsIamPolicy(),
+			"aws_iam_role_policy":              resourceAwsIamRolePolicy(),
+			"aws_iam_role":                     resourceAwsIamRole(),
+			"aws_iam_user_policy":              resourceAwsIamUserPolicy(),
+			"aws_iam_user":                     resourceAwsIamUser(),
```
Comment (bitglue) [inline builtin/providers/aws/provider.go]: @ctiwald here's the list of resources added by this PR

File: builtin/providers/aws/resource_aws_iam_user.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"path": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "/",
+				ForceNew: true,
+			},
```
Comment (bitglue) [inline builtin/providers/aws/resource_aws_iam_user.go]: @ctiwald ...and each of those added resources calls a function like this one, which returns a `schema.Resource` like this. In particular, the `Schema:` map enumerates all of the attributes of that resource. This one as an example has `unique_id`, `name`, and `path`. We can also see from the attributes of each resource that `unique_id` is computed. If it's listed here, it's exposed to the configuration language and should be documented.

Comment (ctiwald) [inline builtin/providers/aws/resource_aws_iam_user.go]: Noted. The docs I've given you therefore need more work. I'll keep chugging along and let you know in [that PR](https://github.com/bitglue/terraform/pull/1) when I think I've got something valid.

File: website/source/docs/providers/aws/r/iam_access_key.html.markdown [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+The following arguments are supported:
+
+* `user` - (Required) The IAM user to associate with this access key.
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The access key ID.
```
Comment (ctiwald) [inline website/source/docs/providers/aws/r/iam_access_key.html.markdown]: One quick question @bitglue -- I don't see `id` in the `Schema:` declaration for `resource_aws_iam_access_key`, but I'm assuming this is discoverable through a different means. Is that the case? Just want to make sure I've got the docs correct for the other resources.

Comment (bitglue) [inline website/source/docs/providers/aws/r/iam_access_key.html.markdown]: Ah yeah...all resources implicitly have an `id`. Grep for `.SetId(` to find it.

**Comment by ctiwald** [comment]
Alright a PR against your branch has everything but iam_role. I can try to knock that out tomorrow or, if you merge that, someone can take it from there.

**Comment by bitglue** [comment]
@mitchellh thoughts on merging this?

File: website/source/docs/providers/aws/r/iam_instance_profile.html.markdown [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+Provides an IAM instance profile.
+
+## Example Usage
+
+```
+resource "aws_iam_role" "role" {
+    name = "test_role"
+    path = "/"
+    policy = <<EOF
```
Comment (ravenac95) [inline website/source/docs/providers/aws/r/iam_instance_profile.html.markdown]: This should be `assume_role_policy` if I'm not mistaken

**Comment by ravenac95** [comment]
I found some issues in testing this on my own aws account. When creating everything from an aws_iam_role, aws_iam_role_policy, aws_iam_instance_profile, and an aws_instance terraform operates too quickly for changes to propagate. So what seems to happen is that aws_iam_instance_profile gets "created" but when aws_instance attempts to spin up an EC2 instance the aws_iam_instance_profile resource is not yet ready.

Essentially I see the following error:

```
Error applying plan:

1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* Error launching source instance: InvalidParameterValue: Value (test-server-profile) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

I may have some time to look at this over the course of this week, but I just wanted to make sure that this problem was known. Just in case someone had some ideas off the top of their head.

**Comment by bitglue** [comment]
@ravenac95 I've never had that happen to me. Do you have a repro case you can share?

**Comment by ravenac95** [comment]
@bitglue here's what I have that consistently breaks. (There's a step by step at the bottom)

In test.tf:

```
variable "aws_access_key" {}
variable "aws_secret_key" {}

variable "ec2_ami" {
    default = "ami-d05e75b8"
}
variable "ec2_key" {}

variable "ec2_instance_type" {
    default = "m3.medium"
}

variable "server_name" {
    default = "foo-server"
}
variable "server_type" {
    default = "foo-server"
}
variable "server_foo_volume_size" {
    default = 1024
}
variable "server_foo_volume_type" {
    default = "standard"
}

provider "aws" {
    access_key = "${var.aws_access_key}"
    secret_key = "${var.aws_secret_key}"
    region = "us-east-1"
}

resource "aws_iam_role" "foo_server" {
    name = "${ var.server_type }-role"

    assume_role_policy = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [{
        "Effect": "Allow",
        "Principal": { "Service": "ec2.amazonaws.com" },
        "Action": "sts:AssumeRole"
    }]
}
EOF
}

resource "aws_iam_role_policy" "foo_server" {
    name = "${ var.server_type }-role-policy"
    role = "${ aws_iam_role.foo_server.name }"
    policy = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [{
        "Action": [
            "ec2:*"
        ],
        "Effect": "Allow",
        "Resource": "*"
    }]
}
EOF
}

resource "aws_iam_instance_profile" "foo_server" {
    name = "${ var.server_type }-profile"
    roles = ["${ aws_iam_role.foo_server.name }"]
}

resource "aws_security_group" "foo_server" {
    name = "${ var.server_name }"
    description = "Security group for foo server"

    ingress {
        from_port = 22
        to_port = 22
        protocol = "tcp"
        cidr_blocks = ["0.0.0.0/0"]
    }
}

resource "aws_instance" "foo_server" {
    ami = "${ var.ec2_ami }"
    instance_type = "${ var.ec2_instance_type }"
    security_groups = ["${ aws_security_group.foo_server.name }"]
    key_name = "${ var.ec2_key }"

    tags {
        Name = "${ var.server_name }"
        Type = "${ var.server_type }"
    }

    ebs_block_device {
        device_name = "/dev/sdf"
        volume_size = "${ var.server_foo_volume_size }"
        volume_type = "${ var.server_foo_volume_type }"
        encrypted = true
        delete_on_termination = false
    }

    iam_instance_profile = "${ aws_iam_instance_profile.foo_server.name }"
}

output "ip" {
    value = "${aws_instance.foo_server.public_dns}"
}
```

Then in a test.tfvars file:

```
aws_access_key = "...key..."
aws_secret_key = "...key..."
ec2_key = "...key-name..."
```

First, let me know if I'm doing something silly here. Perhaps I just misconfigured something.

*Step-by-Step Reproduction*

1.  Run `terraform apply -var-file test.tfvars`
2. That will fail with an error like the one from above except `(foo-server-profile)` instead of `(test-server-profile)`
3. Run `terraform apply -var-file test.tfvars` again
4. That should work!

**Comment by catsby** [comment]
Hey all – how do we feel about the state of this pull request? Are there questions/issues outstanding, or do you feel it's ready to be merged?

**Comment by johnrengelman** [comment]
I think the errors being described above are likely AWS "eventual consistency" type things. So I think we should merge this to get it out there and then fix those types of errors as the come up.

**Comment by bitglue** [comment]
Agree.

**Comment by bitglue** [comment]
@ravenac95 thanks for the repro case, BTW. I can reproduce, but I don't have the bandwidth to fix it now. I thought maybe #1732 would be a way forward. If this gets merged, probably you should move the repro case to a new issue.

**Comment by ravenac95** [comment]
@bitglue No worries. I may poke around to see if there's any way to bring a better solution to AWS's eventual consistency, but for the most part I think the problems encountered by my case are not indicative of a problem in this PR but represent a problem at a higher level. So ya, I agree this should be merged as well :)

**Comment by catsby** [comment]
Hey @bitglue – this looks awesome, thanks!

We do have at least one request, and that is to break out the changes you have made here to schema helper, with the `schema.HashString` method and friends, into a separate pull request. I realize some of your IAM resources use that method, so we'll be sure to merge it first, but it would help separate concerns here. 

Let us know if you can work on that, and thanks again!

**Comment by bitglue** [comment]
@catsby It's already [a separate commit](https://github.com/bitglue/terraform/commit/3756e425c370f65a8fdee7682efd43c3b5ba797c). If you want to merge it separately, I think just `git merge 3756e425c370f65a8fdee7682efd43c3b5ba797c` will do that.

**Comment by bitglue** [comment]
Rebased to fix conflicts, so now that commit is `33183c0`.

**Comment by johnrengelman** [comment]
You should probably just create a new PR from that commit so they can merge it with a button click. Just git the button from here: https://github.com/hashicorp/terraform/compare/master...bitglue:33183c078bfb820d32a0135c5ec928aa88d77c5c

**Comment by bitglue** [comment]
No thanks. Then I'll have *two* PRs to keep coherent and free of conflicts.

**Comment by catsby** [comment]
Pulling this in, thanks.

**Pull request merged**